forge/last-checker.rkt

```diff
@@ -129,11 +218,31 @@
              [else (cons (string->symbol (string-append (symbol->string signame) "_remainder"))
                          all-primitive-descendants)])])))
 
+; wrap around checkExpression-mult to provide check for multiplicity, 
+; while throwing the multiplicity away in output;
+(define (checkExpression run-or-state expr quantvars checker-hash)
+  (let ([output (checkExpression-mult run-or-state expr quantvars checker-hash)])
+    (car output)))
+
+; similar except that this is called from a formula, so in bsl 
+; it should check that the multiplicity of the overall expr is 1 
+(define (checkExpression-top run-or-state expr quantvars checker-hash)
+  (let ([output (checkExpression-mult run-or-state expr quantvars checker-hash)])
+      ; insert check for (first here)
+    ;(printf "checkExpression-top; node: ~a  ; mult: ~a" expr (cdr output))
+    (when (hash-has-key? checker-hash 'expr-mult) ((hash-ref checker-hash 'expr-mult) expr (cdr output)))
+    (car output)))
+
 ; For expressions, this descent does two things:
 ;   - collects possible typings of expressions
 ;   - throws syntax errors as needed (using typing info from children)
-(define/contract (checkExpression run-or-state expr quantvars)
-  (@-> (or/c Run? State? Run-spec?) node/expr? list? (listof (listof symbol?)))
+(define/contract (checkExpression-mult run-or-state expr quantvars checker-hash)
+  (@-> (or/c Run? State? Run-spec?)
+       node/expr?
+       list?
+       hash?
+       any)
+       ;(listof (listof (listof symbol?)))
```

Comment:
why weaken this to `any` ?

[[ Next time, I want to understand what's happening to the `checkExpression` interface with all these cons / car / cdr ]]

Reply:
create some struct for checkExpression return

Reply:
something like "lc-results"